Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp

```diff
@@ -126,144 +123,156 @@ void SubSearcher(const ROMol &in_query, const Bits &bits,
     unsigned int startIdx, unsigned int endIdx, bool recursionPossible,
-    bool useChirality, bool useQueryQueryMatches, int numThreads = -1,
-    int maxResults = 1000) {
+    bool useChirality, bool useQueryQueryMatches,
+    int numThreads = -1, int maxResults = 1000,
+    std::vector<unsigned int> *idxs = nullptr) {
   PRECONDITION(startIdx < mols.size(), "startIdx out of bounds");
   PRECONDITION(endIdx > startIdx, "endIdx > startIdx");
-  numThreads = (int)getNumThreadsToUse(numThreads);
 
-  endIdx = std::min(mols.size(), endIdx);
-  if (endIdx < static_cast<unsigned int>(numThreads)) {
-    numThreads = endIdx;
+  // do not do any work if no results were requested
+  if (maxResults == 0) {
+    return 0;
   }
 
-  std::vector<std::future<void>> thread_group;
-  std::atomic<int> counter(0);
-  std::vector<std::vector<unsigned int>> internal_results(numThreads);
+  endIdx = std::min(mols.size(), endIdx);
+
+  numThreads = static_cast<int>(getNumThreadsToUse(numThreads));
+  numThreads = std::min(numThreads, static_cast<int>(endIdx));
 
   bool needs_rings = query_needs_rings(query);
   Bits bits(fps, query, recursionPossible, useChirality, useQueryQueryMatches);
+  int counter = 0;
 
-  for (int thread_group_idx = 0; thread_group_idx < numThreads;
-       ++thread_group_idx) {
-    // need to use boost::ref otherwise things are passed by value
-    thread_group.emplace_back(
-        std::async(std::launch::async, SubSearcher, std::ref(query), bits,
-                   std::ref(mols), std::ref(internal_results[thread_group_idx]),
-                   startIdx + thread_group_idx, endIdx, numThreads,
-                   std::ref(counter), maxResults, needs_rings));
-  }
-  for (auto &fut : thread_group) {
-    fut.get();
+#ifdef RDK_THREADSAFE_SSS
+  if (numThreads > 1) {
+    std::vector<int> counterVect(numThreads, 0);
+    int maxResultsPerThread = maxResults;
+    if (maxResults != -1) {
+      maxResultsPerThread /= numThreads;
+    }
+    std::vector<int> maxResultsVect(numThreads, maxResultsPerThread);
+    std::vector<unsigned int> endIdxVect(numThreads, endIdx);
+    if (maxResults != -1) {
+      int excess = maxResults % numThreads;
+      for (int i = 0; i < excess; ++i) {
+        ++maxResultsVect[i];
+      }
+    }
+    std::vector<std::future<void>> thread_group;
+    std::vector<std::vector<unsigned int>> internal_results;
+    if (idxs) {
+      internal_results.resize(numThreads);
+    }
+    int thread_group_idx;
+    for (thread_group_idx = 0; thread_group_idx < numThreads;
+         ++thread_group_idx) {
+      // need to use boost::ref otherwise things are passed by value
+      thread_group.emplace_back(
+          std::async(std::launch::async, SubSearcher, std::ref(query), bits,
+                     std::ref(mols), startIdx + thread_group_idx,
+                     std::ref(endIdxVect[thread_group_idx]), numThreads,
+                     needs_rings, std::ref(counterVect[thread_group_idx]),
```

Comment:
We might be able to clean the code up a tad if we introduce a thread barrier and share the counterVect, but that can be a future optimization.